pkg/algorithm/cell_allocation.go

```diff
@@ -0,0 +1,304 @@
+// This function will try to avoid using cells whose resources are not within the suggested nodes if possible.
+// In case there is no cell within the suggested nodes, it will try to randomly return a healthy node with at least
+// one opportunistic pod to try preemption. If that also fails, it just returns the min-opportunistic cell,
+// regardless of if it is suggested or healthy.
+func getFewestOpporPhysicalCell(cl CellList, suggestedNodes common.Set) *PhysicalCell {
+	fewestOpporNum := int32(math.MaxInt32)
+	fewestOpporNumSuggested := int32(math.MaxInt32)
+	var fewestOpporCell *PhysicalCell
+	var fewestOpporCellSuggested *PhysicalCell
+	var preemptibleCells []*PhysicalCell
+	for _, c := range cl {
+		if pc := c.(*PhysicalCell); pc.GetVirtualCell() == nil && pc.GetPreBoundVirtualCell() == nil {
+			opporNum := pc.GetUsedGpuNumAtPriorities()[opportunisticPriority]
+			if opporNum < fewestOpporNum {
+				fewestOpporNum = opporNum
+				fewestOpporCell = pc
+			}
+			allNodesInSuggested := true
+			nodes, _ := pc.GetPhysicalPlacement()
+			for _, n := range nodes {
+				if !suggestedNodes.Contains(n) {
+					allNodesInSuggested = false
+					break
+				}
+			}
+			if allNodesInSuggested && opporNum < fewestOpporNumSuggested {
+				fewestOpporNumSuggested = opporNum
+				fewestOpporCellSuggested = pc
+			}
+			if opporNum > 0 {
+				preemptibleCells = append(preemptibleCells, pc)
+			}
+		}
+	}
+	var selectedCell *PhysicalCell
+	if fewestOpporCellSuggested != nil {
+		selectedCell = fewestOpporCellSuggested
+		nodes, _ := selectedCell.GetPhysicalPlacement()
+		klog.Infof("Selected a cell within suggested nodes: %v, nodes %v", selectedCell.GetAddress(), common.ToJson(nodes))
+	} else if len(preemptibleCells) > 0 {
+		// If we cannot find a cell within suggested nodes, we will try to preempt some pods instead of
+		// directly returning the fewestOpporCell (because this cell could be a bad node, we should not return it).
+		// Also, we will choose a random cell, to avoid always returning the same cell (similar to above,
+		// if we always return the same cell, it might be a bad node, preempting pods on a bad node won't bring
+		// it back to the suggested nodes)
+		selectedCell = preemptibleCells[rand.Int31n(int32(len(preemptibleCells)))]
+		nodes, _ := selectedCell.GetPhysicalPlacement()
+		klog.Infof("Selected a cell not fully within suggested nodes (some of its children may be within); "+
+			"preempting opportunistic pods may help: %v, nodes %v",
+			selectedCell.GetAddress(), common.ToJson(nodes))
+	} else if fewestOpporCell == nil {
+		panic("VC Safety Broken: Cannot find any physical cell that has not been bound to a virtual cell")
+	} else {
+		selectedCell = fewestOpporCell
+		nodes, _ := selectedCell.GetPhysicalPlacement()
+		klog.Infof("Selected a cell not fully within suggested nodes (some of its children may be within); "+
+			"no preemption can help: %v, nodes %v",
+			selectedCell.GetAddress(), common.ToJson(nodes))
+	}
```

Comment:
During filteringPhase, Rack A partial bad nodes, Rack B all bad nodes, here you may select Rack B even if user only ask one node, so hang.

Reply:
Will fix in next release: 
**VC scheduling should aware bad/suggested nodes**

---
In reply to: [399218497](https://github.com/microsoft/hivedscheduler/pull/8#discussion_r399218497) [](ancestors = 399218497)